pm: add deploy workflow validation before merge

## Pattern observed
Weekly scan found repeated deploy failures that escaped review because workflow syntax, Dockerfile/runtime parity, and service-specific deployment contracts were not validated before merge:
- qwickapps/ide#107, #109 and PRs qwickapps/ide#108, #110, #111: Docker build used Node 20/22 while CI used Node 24, and the actual deployed `Dockerfile.ts` was missed on the first fix.
- qwickapps/documents#30 / PR qwickapps/documents#31 and qwickapps/documents#29: pure-Go deploys and security workflow YAML broke because reusable workflow assumptions were not validated against service type.
- qwickapps/qwickwatch#237 / PR qwickapps/qwickwatch#239: production `next build` was missing from PR checks, allowing route-export/build-only failures to reach deploy.
- qwickapps/faabzi#97 / PR qwickapps/faabzi#98: deploy expected a root Dockerfile that the repo did not provide.

## User impact
Routine deploys can fail after merge, block release recovery work, and force manual triage during incidents. The impact is highest for auth, monitoring, and backup services where delayed deploys directly affect operator recovery.

## Suggested investigation
Add a reusable CI preflight that runs before merge and before deploy for each service contract: workflow YAML parse/actionlint, expected Dockerfile path exists, service runtime matches package manager requirements, required build command runs in production mode, and service-type toggles such as `dockerfile_only` are validated. Publish the contract as a small manifest per repo so service-specific exceptions are explicit instead of discovered during deployment.

## Test report — deploy validation gaps

**Scope**: ci-workflows#25 — YAML syntax, Dockerfile/runtime assumptions, missing production builds, service contract mismatches  
**Services**: ide, documents, qwickwatch, faabzi  
**Suite run**: manual evidence-gathering from run logs + repo state — no automated suite to run against ci-workflows#25 itself (the preflight workflow doesn't exist yet)

---

### Current deploy status (as of 2026-05-25)

| Service | Last deploy | Status | Note |
|---------|------------|--------|------|
| ide | run 26403666439 @ 13:45 | ❌ FAILING | New failure: pnpm esbuild ignored-builds (see G8 below) |
| documents | no deploy runs today | — | No recent trigger; last known good after #31 |
| qwickwatch | run today @ 13:20 | ✅ PASS | CI + deploy both green after #239 |
| faabzi | run 26410939738 @ 16:48 | ❌ FAILING | GHCR 403 on push — 5 consecutive failures |

---

### Coverage assessment

#### Gaps not covered by pre-merge checks (still open)

**G1 — No actionlint gate on consumer-repo PRs**  
`ci-workflows` has `.github/actionlint.yaml` but there is no step in the reusable `pr-validate.yml` that runs actionlint on the calling repo's workflow files. The documents#26/29 YAML parse error (multiline `run:` block scalar indentation) would survive PR review and be caught only at runtime.  
_Evidence_: `pr-validate.yml` Gate 1-5 — no actionlint invocation anywhere.

**G2 — No Dockerfile path existence pre-check**  
`pr-validate.yml` Gate 3 builds the Dockerfile only if `has_dockerfile=true`, but there is no earlier step that asserts the file exists at `inputs.dockerfile_path`. A caller can set `dockerfile_path: Dockerfile.ts` but CI passes `Dockerfile` and the mismatch is invisible until deploy.  
_Evidence_: ide needed three PRs (#108 → #110 → #111) because the actually-deployed file (`Dockerfile.ts`, used via `-f ./Dockerfile.ts` in `deploy.yml`) was separate from the first-fixed `Dockerfile`. Gate 3 would have validated the wrong file.

**G3 — No Dockerfile FROM version vs. CI runtime parity check**  
`pr-validate.yml` accepts `node_version` as an input (default `'20'`) but never extracts the `FROM node:XX` version from the Dockerfile and compares them. Callers can set `node_version: '24'` in CI but leave `FROM node:22-alpine` in the Dockerfile.  
_Evidence_: ide#107 (node:20 vs node:24), ide#109 (node:22 insufficient — same class). The second fix required knowing that pnpm 11 corepack specifically needs node:24, not just 22+.

**G4 — No service-type vs. `dockerfile_only` consistency check**  
`deploy-app.yml` provides a `dockerfile_only` toggle to skip pnpm/corepack setup for non-TS services, but there is no preflight that validates: if the Dockerfile's first `FROM` is `golang:*`, `dockerfile_only` must be `true`. Callers must discover the mismatch at deploy time.  
_Evidence_: documents#30 — default `dockerfile_only: false` triggered corepack on a pure-Go service (`FROM golang:1.21-alpine`).

**G5 — No per-repo service contract manifest**  
Service-specific constraints (expected Dockerfile path, runtime, `dockerfile_only`, required build command) are embedded implicitly across multiple workflow files in each repo. There is no machine-readable manifest a preflight script can validate against.  
_Evidence_: Every incident in this issue (ide#107-111, documents#30, qwickwatch#237, faabzi#97) was discoverable from static analysis had a manifest existed.

**G6 — Production build gate is per-repo, not enforced by the reusable workflow**  
`pr-validate.yml` has a `build_command` input but it defaults to empty (skipped). There is no mechanism that requires Next.js services to pass `build_command: 'pnpm build'`. The qwickwatch fix (PR #239) added `next build` to `ci.yml` directly, so a future Next.js service onboarding via `pr-validate.yml` will silently skip the production build unless the caller explicitly passes the input.  
_Evidence_: qwickwatch#237 — build-only errors (route export type violations) reached deploy because no production build ran in CI.

---

### Active failures requiring immediate attention (not blocked on ci-workflows#25)

**G7 — faabzi GHCR push: 403 Forbidden (5 consecutive failures, blocked deploy)**  
- Error: `failed to push ghcr.io/qwickapps/img-faabzi-dev: 403 Forbidden` at HEAD request to blob endpoint  
- Root cause: GHCR package `qwickapps/img-faabzi-dev` has never been successfully pushed; the push token lacks `write:packages` permission or the package visibility needs initializing  
- The Dockerfile builds successfully (placeholder node:22-alpine from PR #98)  
- Additional note: faabzi Dockerfile uses `node:22-alpine` — will hit the same node:sqlite/corepack gap that hit ide if pnpm 11 is ever added  
- **Action needed**: ops to verify GHCR token scope for `img-faabzi-dev` and initialize the package  
- Tracked separately from ci-workflows#25 (ops issue, not a preflight gap)

**G8 — ide deploy: `ERR_PNPM_IGNORED_BUILDS` for esbuild (today's failure)**  
- Error: `Ignored build scripts: esbuild@0.27.4` → `pnpm install --frozen-lockfile` returns exit 1 inside Docker build  
- Root cause: pnpm 11's default security settings block build scripts for native packages (esbuild, etc.) unless explicitly approved via `pnpm approve-builds` or `.pnpmfile.cjs`  
- The node:24-alpine fix from PR #111 resolved the `node:sqlite` error but exposed this new pnpm 11 strictness  
- This is the same class as G3: a pnpm behavior gap between CI (which has approved builds on the macmini runner) and Docker (fresh environment with default pnpm config)  
- **Action needed**: ide coder to add `pnpm config set allow-build esbuild` (or equivalent `onlyBuiltDependencies` in package.json) and update Dockerfile  
- Tracked separately: recommend filing ide#112

---

### Tests added in this PR session

None — ci-workflows#25 is a new-feature request, not an existing test. The preflight reusable workflow does not yet exist.

---

### Tests recommended (to be implemented in ci-workflows)

The following test cases should be written as part of the preflight implementation:

**T1 — actionlint gate (addresses G1)**  
- `actionlint .github/workflows/*.yml` passes on well-formed YAML  
- Fails on the documents#26 pattern (unindented shell body in multiline `run:`)  
- Fails on a missing `uses:` version pin  

**T2 — Dockerfile path existence preflight (addresses G2)**  
- For each `dockerfile_path` value referenced in any deploy workflow in the calling repo, assert the file exists  
- Fails if `Dockerfile.ts` is referenced in deploy but only `Dockerfile` exists  
- Passes when both files exist  

**T3 — FROM version vs CI runtime parity (addresses G3)**  
- Extract `FROM node:XX-alpine` from each Dockerfile  
- Assert it equals the `node_version` or `NODE_VERSION` env var in the deploy workflow  
- Fails on node:20/node:22 vs node:24 mismatch  
- Skip check for multi-stage builds where runner stage FROM is intentionally different  

**T4 — service type vs. `dockerfile_only` consistency (addresses G4)**  
- If any Dockerfile's first non-ARG `FROM` is `golang:*`, assert caller's `dockerfile_only: true`  
- Fails on Go service with `dockerfile_only` absent or false  
- Passes on TS services or Go services with explicit `dockerfile_only: true`  

**T5 — Next.js service production build present (addresses G6)**  
- If `package.json` contains `next` in `dependencies` or `devDependencies`, assert either:  
  - `build_command` passed to `pr-validate.yml` contains `next build` / `pnpm build`  
  - OR a dedicated `build` job exists in `ci.yml` running `next build`  
- Fails when neither is present  
- qwickwatch passes post-#239; a fresh service without the build job would fail  

**T6 — GHCR package write pre-check (addresses G7 class)**  
- Before `docker buildx build --push`, assert the push token can write to the target GHCR package  
- Method: `curl -H "Authorization: Bearer $TOKEN" https://ghcr.io/v2/qwickapps/<pkg>/tags/list` → 200 or 404 (not 403)  
- 403 at pre-check fails fast with a meaningful error instead of after a full image build  

---

### Verdict

- [ ] Behavior matches PR claim — **no preflight workflow exists yet; this is the pre-implementation audit**
- [ ] No regressions detected — **two active regressions confirmed** (ide G8, faabzi G7)
- [ ] Coverage adequate for merge — **not applicable; nothing to merge yet**

**Summary**: Six structural gaps confirmed across the four services. Two active deploy failures (ide esbuild pnpm 11 strictness, faabzi GHCR 403) require immediate ops/coder action independent of ci-workflows#25. The gap inventory above and the six test cases (T1–T6) constitute the acceptance criteria for the preflight reusable workflow requested in this issue.

--- a/.github/workflows/pr-validate.yml
+++ b/.github/workflows/pr-validate.yml
@@ -2,8 +2,8 @@ name: PR Validate
 
 # Reusable PR validation workflow for QwickApps repositories.
 #
-# Provides 6 sequential gates:
-#   1. validate-commits  — attribution check + branch target
+# Provides 7 sequential gates:
+#   1. validate-commits  — attribution check + branch target + actionlint workflow lint
 #   2a. lint-and-typecheck — tsc --noEmit (TS) or go vet + golangci-lint (Go)
 #   2b. test              — pnpm test (TS) or go test ./... (Go)
 #   3. build              — optional build command + Docker image build
@@ -65,6 +65,7 @@ on:
 # GATE 1: validate-commits
 #   • Branch target: PRs from feature branches must not target 'main' directly
 #   • Attribution:   No AI co-authorship trailers permitted (via attribution-check.sh)
+#   • Actionlint:    Validate all workflow files in .github/workflows
 # ─────────────────────────────────────────────────────────────────────────────
 jobs:
   validate-commits:
@@ -113,6 +114,15 @@ jobs:
         run: |
           chmod +x .ci-workflows/scripts/attribution-check.sh
           .ci-workflows/scripts/attribution-check.sh
+
+      - name: Lint workflow files (actionlint)
+        run: |
+          export PATH="/opt/homebrew/bin:/usr/local/bin:/usr/bin:/bin:$PATH"
+          if [ ! -d ".github/workflows" ]; then
+            echo "::notice::No .github/workflows directory found — skipping actionlint"
+            exit 0
+          fi
+          actionlint .github/workflows/*.yml
 
 # ─────────────────────────────────────────────────────────────────────────────
 # GATE 2a: lint-and-typecheck  (needs: validate-commits)

